Add GAME ID send for VMU PRO for Per Game VMU functionality

nexus382 opened the pull request:
Adds OpenMenu-style GameID support for Dream Pico Port.
On game start or state load, Flycast now forwards "settings.content.gameId" through DreamLink, and the DPP backend sends a VM2/VMU Pro–compatible command 33 packet (memcard func + 12-byte ID) to the VMU slot. This makes VMU Pro/VM2 auto-select per-game storage in Flycast the same way they already do when launched via OpenMenu on real hardware so Flycast can also enjoy the functionality as well..

nexus382 committed:
Send game ID when starting game or loading state

ensure we send the GAME ID on both load states and boot to ensure this is sent any way a game is run.

nexus382 committed:
Add sendGameId method to SDLGamepad interface

Added a method to send the current game ID to a DreamLink device.

nexus382 committed:
Add sendGameId method for game ID transmission

followed OpenMenu code to see exactly what is sent to the vmupro/vm2,  Now, "sendGameId" will mirror OpenMenu’s VMU Pro/VM2 protocol by issuing Maple command 33 with a storage function word, 12-byte GameID payload, and 4-word length to the controller expansion address

nexus382 committed:
Add names for Dreamcast peripherals

nexus382 committed:
Update dreampicoport.cpp

Fixed sent value

nexus382 committed:
Fix memcpy offset and message size in dreampicoport

msg.size changed to 3, and msg.data from 4 to 0

nexus382 committed:
Adjust message data copy and size for game ID

testing

Tails86 commented on core/hw/maple/maple_devs.cpp:
```diff
@@ -2628,6 +2629,11 @@ void createDreamLinkDevices(std::shared_ptr<DreamLink> dreamlink, bool gameStart
 					dreamLinkVmus[i].push_back(vmu);
 				}
+
+				if (gameStart || stateLoaded)
```
`stateLoaded` isn't needed. The device will disconnect on `stateLoaded` anyway because it isn't compatible with flycast

nexus382 replied:
I see your point, it was a future proof, but thinking more about it, youd have to start the game before you load state anyway, so yes this may be removed. good ee sir.

Tails86 commented on core/hw/maple/maple_devs.cpp:
```diff
@@ -2628,6 +2629,11 @@ void createDreamLinkDevices(std::shared_ptr<DreamLink> dreamlink, bool gameStart
 					dreamLinkVmus[i].push_back(vmu);
 				}
+
+				if (gameStart || stateLoaded)
```
Should be checking if `vmu.useRealVmuMemory` is true before doing this as well

Tails86 commented on core/hw/maple/maple_devs.cpp:
```diff
@@ -2628,6 +2629,11 @@ void createDreamLinkDevices(std::shared_ptr<DreamLink> dreamlink, bool gameStart
+
+				if (gameStart || stateLoaded)
+				{
+					dreamlink->sendGameId(i, settings.content.gameId);
```
Should probably call `vmu.OnSetup()` before this call to ensure any cached data is cleared

nexus382 committed:
Call OnSetup before Setup for VMU

the `OnSetup` call sits right before `Setup` inside `createDreamLinkDevices` because that is the point where we are about to rebuild or update the DreamLink VMU.so onw as  the call is happening due to a game start, or first-time creation. Clearing cached data right here guarantees that every following `Setup` starts from a clean slate, f whether we reused an existing VMU object or just created one.

nexus382 committed:
Fix VMU setup and game start conditions

nexus382 committed:
Fix spaceing

nexus382 committed:
spaces

Tails86 reviewed (approved):
This is using the game ID command specifically for the VmuPro: https://github.com/megavolt85/openmenu/blob/master/vm2_api.c

This should be benign to a normal VMU.

nexus382 committed:
Clean up on sanity checking code

nexus382 committed:
Sanity Checking for Clean code work.

nexus382 committed:
simplify Game ID memcpy and drop (std::min) workaround

copy gameId directly into Maplemsg data with a single memcpy relying on zero-initialized msg for padding

remove unnecessary parentheses around std::min now that NOMINMAX is defined globally

Tails86 committed:
Minor bug fixes for DreamPicoPort

nexus382 committed:
rge remote-tracking branch 'upstream/dev' into nexus382-Per-Game-ID-for-VMU-PRO/VM2

nexus382 committed:
Merge Tails86 minor dream pico port corrections
